### DISCUSSION TITLE
Welcome to mongock Discussions!

### DISCUSSION THREAD

**Discussion by dieppa** [flamingock/mongock#56] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by ghost** [comment]
Hello, does mongock support multiple databases configuration? Can i have two databases -> 2 mongo templates, and than in changelog choose which database run changelogs on? Can i create multiple mongocksTemplate? Thanks

**Reply by dieppa** [reply]
Hello @sebaszczen , the quick answer is no. And I struggle to think a good scenario where this is needed. keep in mind that the lock and changelogs are persisted in database, so Mongock needs to know which database to write to. 

Something totally difference is to have two independent Mongock runners using different database each, in that case each Mongock instance knows in advance which database to write to and that will possible. Although right now there is no a clear support for that, you can create two separated  beans, injecting to each one the right MongoTemplate.

What you never should be doing is to run two different Mongock instances pointing to the same database. That would produce unexpected and undesired results.

**Reply by ghost** [reply]
@dieppa, hey thank you for the response. If i have two  independent Mongock runners how can i use one? "you can create two separated beans" what type of bean do you mean? 
Right now i define changeSet that way:

`@ChangeLog(order = "002")
public class Xxx {

    @ChangeSet(order = "0000", id = "xx", author = "xx", runAlways = true)
    public void changeLog0000(MongockTemplate mongockTemplate) {`
How is that MongoTemplate related to those Mongock runners? How can i choose one?

**Reply by dieppa** [reply]
Hello, in our road map for the version 5, for which we should have a RC at the end June - middle of July, we are providing two features related to this:

## Multitenant 
**Description:** This feature allows you to have a single mongock runner with multiple database to perform your changelogs. Ideal for consultancies that provide SaaS application for multiple users, each of them with their own database, but all of them require to perform the migrations.

**Development status** Very advanced. In testing stage

You can find more information in this [discussion](https://github.com/cloudyrock/mongock/discussions/331)


## Multidatabase
**Description** This feature allows you to connect to multiple databases in a single changeLog. Te purpose is to have a target database to perform the migration(which is synchronised with the lock) and other source database to retrieve potential information

**Development status** Finished and tested. We can provide a beta version for testing purpose

You can find more information in this [discussion](https://github.com/cloudyrock/mongock/discussions/334)

**Reply by ghost** [reply]
@dieppa thanks for the response. So is it impossible to reach my goal so far? I would like to have 2 databases and than perform some changelogs against one database, and some changelogs against the other one.

**Reply by dieppa** [reply]
Then, probably the best solution is to use two mongock instances. For that you need to use the builder approach. Please see our [documentation](https://www.mongock.io/quick-start#traditional-builder-approach).

Then you can create two different named MongockApplicationRunner or beans and spring will run then:
```java
@Bean("mongock-runner1")
public MongockApplicationRunner mongockApplicationRunner(
        ApplicationContext springContext,
        MongoTemplate mongoTemplate1) {
    return MongockSpring5.builder()
        .setDriver(SpringDataMongo3Driver.withDefaultLock(mongoTemplate1))
        .addChangeLogsScanPackage("your_changeLog_package_path")
        .setSpringContext(springContext)
        .buildApplicationRunner();
  }
  //...
  @Bean("mongock-runner2")
public MongockApplicationRunner mongockApplicationRunner(
        ApplicationContext springContext,
        MongoTemplate mongoTemplate2) {
    return MongockSpring5.builder()
        .setDriver(SpringDataMongo3Driver.withDefaultLock(mongoTemplate2))
        .addChangeLogsScanPackage("your_changeLog_package_path")
        .setSpringContext(springContext)
        .buildApplicationRunner();
  }

```

Please note that they use different MongoTemplate and therefor, different databases. Then the MongockTemplate injected will be the one based on each MongoTemplate
